apis/v1beta1/gateway_types.go

```diff
@@ -450,6 +467,8 @@ type RouteGroupKind struct {
 }
 
 // GatewayAddress describes an address that can be bound to a Gateway.
+//
+// +kubebuilder:validation:XValidation:message="Hostname value must only contain valid characters (matching ^(\\*\\.)?[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$)",rule="self.type == 'Hostname' ? self.value.matches('^(\\\\*\\\\.)?[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$'): true"
```

Comment:
FYI: @jpbetz @alexzielenski Two thoughts:

1) This feels like a good candidate for a validation library function, e.g. "IsWildcardDNSDomain" or something.

2) For places that legit need to do a regex match, can we find a way not to repeat the regex twice (and with different levels of escape) ?

Reply:
> 1. This feels like a good candidate for a validation library function, e.g. "IsWildcardDNSDomain" or something.

I agree. As a general rule, I'd like CEL to have full parity with the formats we support, so when we add a format for wildcardDNSDomain, I'd like it to also be possible to validate that in CEL.

> 2. For places that legit need to do a regex match, can we find a way not to repeat the regex twice (and with different levels of escape) ?

We added variable support to ValidatingAdmissionPolicy but not CRD validation rules (with CEL). It would solve for this at the cost of added complexity.  I wonder if having named formats would be sufficient here?  We could document them well and then just reference them by name in the expression and message?